Adds the e2e test suite framework and a bit over 80 test scenarios

iajrz opened the pull request:
This PR has both Go code and a cucumber test suite in gherkin; I'd expect @Olshansk to look at the go side of the matter and @emedrano2112 and @AneudyM to look at the gherkin side of the matter.

Fixes #1482 

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

iajrz committed:
introducing e2e automated tests

iajrz committed:
waiting for server and qol

adds the step "the user waits for n blocks", makes blocks shorter in the configuration (45 seconds now), makes server verbosity optional

iajrz committed:
allow using invalid addresses running commands against server

iajrz committed:
adds 80 something scenarios to the test suite

iajrz committed:
pocket client respects verbosity when running commands

Olshansk commented on e2e/launcher/configuration.go:
```diff
@@ -0,0 +1,52 @@
+	ConfigPath string `json:"config_path"`
+}
+
+const networkConfigurationFileName = "network.json"
```
NIT: `defaultNetworkConfigurationFileName`?

Olshansk replied:
I see a 😄 emoji but no comments, resolution or change. Will you make the change?

iajrz replied:
The emoji was because this literally made me smile and to remind self that I'd seen it. In this case: network configuration file name is not configurable, so it's not "default", but the "actual" name. It doesn't gel to make it configurable, either, so I won't change it.

Olshansk commented on e2e/launcher/configuration.go:
```diff
@@ -0,0 +1,52 @@
+
+const networkConfigurationFileName = "network.json"
+
+func loadNetworkConfiguration(networkConfigDirectory string) (*NetworkConfiguration, error) {
```
```suggestion
func loadNetworkConfiguration(networkConfigDirectory string) (networkConfiguration *NetworkConfiguration, err error) {
```

Given that you have to declare these individually, consider using named return values

Olshansk replied:
I see an 👀 , but no comments/resolution/change. What does this mean?

Olshansk commented on e2e/launcher/configuration.go:
```diff
@@ -0,0 +1,52 @@
+func loadNetworkConfiguration(networkConfigDirectory string) (*NetworkConfiguration, error) {
+	var err error
+	networkConfigDirectory, err = filepath.Abs(networkConfigDirectory)
+
```
```suggestion
```

Olshansk commented on e2e/launcher/network.go:
```diff
@@ -0,0 +1,43 @@
+	networkConfiguration, err := loadNetworkConfiguration("../launcher/network_configs/" + networkConfigDirectory)
+	if err != nil {
+		return
+	}
```
You have pretty nice spaces after err if structures `e2e/launcher/configuration.go` but not here.

I personally prefer the former, but w/e you choose, please stay consistent.

<img width="872" alt="Screenshot 2022-12-13 at 9 43 53 PM" src="https://user-images.githubusercontent.com/1892194/207515753-5ec29b92-b384-4b76-a68a-74e0cee7ab69.png">

iajrz replied:
My personal preferences is the use of spaces -- in fact, blank line before an `else` too when I'm being especially mindful.
Usage in golang seems to go against whitespace. In fact, standard library has no breathing room _within_ functions, although it has plenty _between_ functions.

I see you like the spaces, too. Bad thing about spaces is you need to think of where to put them and configure the editor so you never forget. If you never have space you never have to do that.

How're you doing it in v1 code? How'd you like it to be in the future? I could do either and I'm going to configure the editor to stick to the official option as closely as possible.

Olshansk replied:
Use the spaces approach. Apply everywhere in this PR. It's what we're aiming for in v1 as well.

Olshansk commented on e2e/launcher/network.go:
```diff
@@ -0,0 +1,43 @@
+	log.Printf("root network directory: %v", networkRootDirectory)
+	var nodes []*Node
+	for _, nodeConfiguration := range networkConfiguration.NodeConfigurations {
+		var node *Node
+		node, err = newNode(nodeConfiguration, networkRootDirectory, networkConfiguration.GenesisPath, executablePath)
```
```suggestion
		node, err := newNode(nodeConfiguration, networkRootDirectory, networkConfiguration.GenesisPath, executablePath)
```

iajrz replied:
This is a no-no: in the worst case scenario changing code to be like this would shadow the return value. In this particular case, it would merely cause a compilation error.
Here, two examples:
- [here](https://go.dev/play/p/399zRdoXppd) for the shadowing case (run and look at the logs).
- [here](https://go.dev/play/p/86efKsyifB7) for the compilation error case

Olshansk commented on e2e/launcher/network.go:
```diff
@@ -0,0 +1,43 @@
+	if err != nil {
+		return
+	}
+	log.Printf("root network directory: %v", networkRootDirectory)
```
You use `log.Println("string" + var)` above, but `log.Printf` here. 

What's your choice on when to use which?

iajrz replied:
The code was not written sequentially; I did not stop to think very closely about fmt vs log, and `f` vs `ln`. I will think about this during the revision to comply with the review, though.

Olshansk commented on e2e/launcher/network.go:
```diff
@@ -0,0 +1,43 @@
+		}
+		nodes = append(nodes, node)
+	}
+	network = Network{
+		Nodes:                nodes,
+		NetworkConfiguration: networkConfiguration,
+	}
+	return
```
```suggestion
	return &Network{
		Nodes:                nodes,
		NetworkConfiguration: networkConfiguration,
	}, nil
```

iajrz replied:
Will think about this and address :) both the named returns and the matter of the pointer

iajrz replied:
I don't fancy the multi-line first value and single-line second value. I adopted the pointer, though. Might want to do away with the named returns. Have to think more about _that_.

Olshansk commented on e2e/launcher/network.go:
```diff
@@ -0,0 +1,43 @@
+	NetworkConfiguration *NetworkConfiguration
+}
+
+func LaunchNetwork(networkConfigDirectory string, executablePath string) (network Network, err error) {
```
Why not return a `*Network`?

iajrz replied:
Decision was possibly made because members of the struct already point at things. I will revisit this line of thought and possibly be better able to articulate why I decide whatever I end up with.

Olshansk commented on e2e/launcher/node.go:
```diff
@@ -0,0 +1,147 @@
+	}
+
+	// initialize returnable node
+	node = &Node{
```
In these types of situations, consider `return &Node{...}, nil`

iajrz replied:
I have to think about named vs unnamed returns, which is related to this.

iajrz replied:
Decided against -- the first element would be a multi-line initialization, while the second element (the nil) would get lost lines below the return.

Olshansk commented on e2e/launcher/node.go:
```diff
@@ -0,0 +1,147 @@
+	if err = os.Mkdir(filepath.Join(nodeDataDir, "config"), os.ModePerm); err != nil {
+		return
+	}
+
```
Personal preference (feel free to ignore).

I would've probably declared interim variables for readability and used in place.

```go
configFilePath := filepath.Join(nodeDataDir, "config", "config.json")
genesisFilePath := filepath.Join(nodeDataDir, "config", "genesis.json")
```

iajrz replied:
I generally don't do that except for things that are repeated. Usually 3rd repetition is when I name the thing. Bearing in mind.

Olshansk commented on e2e/launcher/node.go:
```diff
@@ -0,0 +1,147 @@
+}
+
+func writePrivValState(datadir, privValState string) error {
+	cdc := app.Codec()
+	pvkBz, err := cdc.MarshalJSONIndent(privval.FilePVLastSignState{}, "", "  ")
```
Doesn't cost much to inline here. DItto below.

```suggestion
	pvkBz, err := app.Codec().MarshalJSONIndent(privval.FilePVLastSignState{}, "", "  ")
```

iajrz replied:
Indeed, this was a mistake borne from rewriting. Possibly the value of `app.Codec` should live in a package level variable.

Olshansk replied:
No need to leave comments on stuff like this moving forward, if you're okay with the change -> make change -> `Resolve conversation` -> continue

Olshansk commented on e2e/launcher/pocketclient.go:
```diff
@@ -0,0 +1,55 @@
+
+type writerByteArray []byte
+
+func (so *writerByteArray) Write(p []byte) (n int, err error) {
```
s/so/ba so its general?

iajrz replied:
Looks like it should be like that; I think this was lost in translation from where the type had another name.

Olshansk replied:
In the future, no need to comment on stuff like this. Just make the change and `Resolve conversation`.

Olshansk commented on e2e/launcher/pocketclient.go:
```diff
@@ -0,0 +1,55 @@
+	RunCommand(...string) (*CommandResult, error)
+}
+
+func NewPocketClient(executableLocation string, verbose bool) PocketClient {
```
NIT: put factory function below type defitnion.

iajrz replied:
NOT (see what I did there?): this is in the style of the go standard library. I say more about that [in this comment](https://github.com/pokt-network/pocket-core/pull/1514#discussion_r1048961345) -- which if you're reading top to bottom you already saw.

Olshansk commented on e2e/launcher/pocketclient.go:
```diff
@@ -0,0 +1,55 @@
+
+func (pc *pocketClient) RunCommand(commandAndArgs ...string) (*CommandResult, error) {
+	if pc.verbose {
+		fmt.Printf("Running Command: %v\n", commandAndArgs)
```
Why are we using `fmt` instead of `log` everywhere?

iajrz replied:
It's a mistake, as mentioned elsewhere; will think through & normalize. In general I think "useful when using" vs "useful when debugging" should guide that decision, but that might not be what I'm actually doing here.

Olshansk replied:
Let's change to `log`.

Olshansk commented on e2e/launcher/pocketclient.go:
```diff
@@ -0,0 +1,55 @@
+	err := cmd.Run()
+
+	return &CommandResult{
+		Stdout: string(*so),
```
Since we're using the byte array to keep appending and convert it to a string in the end, thoughts on using the strings builder in golang? It implements `Write` by default: https://pkg.go.dev/strings#Builder.Write which is the interface `Stdout`  above needs.

iajrz replied:
Been looking into this; considered a byteBuffer, too. Keeping it in consideration.

Olshansk commented on e2e/launcher/pocketserver.go:
```diff
@@ -0,0 +1,124 @@
+	return nil
+}
+
+type PatternActor interface {
```
DISCUSS: I've always been a fan of declaring all the types in one place or at the top of the file.

I've never gotten accustomed to having types be defined midfile, but is this a common golang pattern?

iajrz replied:
My pattern is shaped by years of Java development (declaration and definition are together, so types end up mid-file). The pattern you prefer reminds me of C and C++ development, where declaration and definition are separate. 

Navigation between files is usually done by symbol cross reference, so we don't waste much time finding things because they're far from the top; furthermore, having all of the types at the top will yield a group of functions unrelated to each other hanging out together at the end of the file.

I had way more text, but decided to go to `io/io.go` in the standard library for guidance. This is how that file is ordered:
- Package level constants and variables
- Interfaces
- Functions with no receivers (operate on interfaces)
- all the structs with their related data around:
  - struct initializer (NewXxYy and similar)
  - struct
  - functions with the struct or pointer to the struct as a receiver

structs are ordered so that the ones below depend on the ones above.

I think this works because of this:
- First you see package level things you might need to customize or refer to
- Then interfaces you might want to implement
- Then functions you might need to use on the interfaces
- Then structs and related implementation nearby
- In the only case you'd read from top to bottom (generally when studying a module to learn it/from it) the fact that types go from more general to more specific means that you get a sense for the architecture / rationale.

I think we should emulate this pattern in our codebases.

Olshansk replied:
Sounds good to me. Leave as is.

Olshansk commented on e2e/launcher/pocketserver.go:
```diff
@@ -0,0 +1,124 @@
+}
+
+type pocketServer struct {
+	executableLocation string
```
By location, do we mean `Path`? If so, could/should we rename it?

iajrz replied:
yes to all

Olshansk replied:
Please update

Olshansk commented on e2e/launcher/pocketserver.go:
```diff
@@ -0,0 +1,124 @@
+type PocketServer interface {
+	Start(...string) error
+	Kill() error
+	RegisterPatternActor(patternActor PatternActor, stream Stream) error
```
DISCUSS: I'm familiar with the "Actor Model" from prior work but this is the first I've heard of "Pattern Actor". 

Can you link to any other work that references this naming convention?

iajrz replied:
Reference to the pattern-action paradigm as seen in awk and similar tools, given the similar usage: if the pattern is matched, we do stuff.
Open to renaming.

Olshansk replied:
Let's rename to `patternAction`

Olshansk reviewed (changes requested):
@iajrz I've still got a long way to go (see screenshot) but figured I'd publish what I have so far so we can start working in parallel as I continue.

<img width="1459" alt="Screenshot 2022-12-13 at 10 10 45 PM" src="https://user-images.githubusercontent.com/1892194/207519813-e569525b-ced1-44e9-927a-24a00f7298ca.png">

iajrz committed:
Update e2e/launcher/configuration.go

spacing consistency

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

Olshansk commented:
@iajrz I'll continue reviewing once the existing comments are resolved or Once the existing comments are resolved or tended

AneudyM commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	And the user should be able to see standard output containing "address"
+
+  Scenario: To query an existing account, wrong address
+	Given the user is running the network "single_node_network"
```
I recommend the use of hooks here to take advantage of reusability. Lots of repeated steps. Background would work for this case.

iajrz replied:
I agree that some scenarios could have fewer visible steps by using Background; as mentioned in the PR's initial comment, though, we're looking for accuracy in the gherkin side, so I think this can stay as is.

AneudyM commented on e2e/tests/accounts.feature:
```diff
@@ -0,0 +1,78 @@
+	Given the user has a pocket client
+	When the user runs the command "accounts sign 7ab712998671b09e1a266ce6901000acb657833b 'example_message'"
+	Then the user should be able to see standard output containing "invalid byte"
+	And the pocket client should have exited without error
```
Ideally, you'd want Then to be the last step, since its purpose is to observe outcomes. Here, each test is validating that the client exited without errors. That should be an individual scenario, not one to do in every scenario. For example:

Given the user has a pocket client
When the user runs the command ...
Then the user should see ...       <-- this is the outcome
---END---

iajrz replied:
I understand where you're coming from; in this case -- a command line interface's API is comprised of its stdout, stderr and exit status. In particular, scripts often use the exit status to know whether the script can continue or needs to abort.

With regards to the `Then...` `And...`, the preference is aesthetic in this case; we're checking for two different outcomes, and it reads easier than `Then...` `Then...`.

iajrz committed:
addressing comment about spacing in error check blocks

iajrz committed:
renaming PatternActor to PatternAction

iajrz committed:
logs now use log, and printf is preferred when building output.

iajrz committed:
making the network live as a pointer instead of a value

iajrz committed:
making return values unnamed

I am pretty uncomfortable about the semantics of this code; I end up declaring a ton of `err` variables for no reason other than they are the second parameter, and the first one needs to be created.

The tendency towards being terse is fighting the common code structuring patterns and the semantics in a royal rumble where the winner gets to still not be satisfactory, so I'll just keep this in a recognizable style and  call it quits until a guiding principle can show up and help us never face such mixed solutions of sometimes creating and sometimes reusing a variable.

iajrz committed:
using strings.Builder instead of custom Writer

opted for it over bytes.Buffer since it seems to be the intended upgrade path anyhow (see https://groups.google.com/g/golang-codereviews/c/6S3_B4N197E?pli=1); they seem similar in the intended use, but it never hurts to use the intended tool.
It also has some checks around escape analysis that were missing from the custom implementation, so we're winning.

iajrz committed:
renamed executableLocation to executablePath

Olshansk commented:
@iajrz I saw that you pushed some updates. Please re-request a review (see screenshot) when it's ready for another round

![Screenshot 2022-12-20 at 3 50 38 PM](https://user-images.githubusercontent.com/1892194/208763831-4f42ed9c-d799-4ecf-8f0a-95dabbd7d2bb.png)

emedrano2112 commented on e2e/tests/accounts.feature:
```diff
@@ -0,0 +1,78 @@
+
+  Scenario Outline: Running Commands With Correct Syntax and Bad Data
+	Given the user has a pocket client
+	Given the user runs the command <bad_command>
```
Change given, by When. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

iajrz replied:
These checks don't help accuracy. I will go through and revise them, but the accuracy of the test suite won't change any. The gherkin revision is meant to help determine whether the tests are accurate -- ie: sufficient to guarantee that the behavior they seem to be testing is actually correct.

ivan-pokt replied:
seconded! We want to have no more than one Given statement for every scenario

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing account
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing account, wrong address
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: Pocket Query Account address is invalid But Has valid Bytes
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing account, wrong height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an non-existing Account
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing account txs
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing account txs, wrong address
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query current acl
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario Outline: Unsuccessful Query of an App
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing app parameters
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario Outline: Querying App Parameters, Invalid Height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query the list of existing apps
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an address balance in the network
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an address balance in the network, wrong address
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an address balance in the network, wrong height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query a block at height, wrong height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query a block's at height, given an invalid number height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query a block's at height, given an invalid char in height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query block-txs at height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query a block-txs at height, incomplete
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query the chains height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing node from address
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing node from address, wrong address with only valid bytes
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing node claims
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing node parameters
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query an existing node parameters, wrong address
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query the list of existing nodes
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query the list of existing nodes, Returns error when given wrong height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query the list of existing nodes
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query single param
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query the list of params
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query the list of pocketcore params
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query the list of pocketcore params, Return error with incorrect height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query a specific node's signing info
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To get supplies
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To get supplies, Returns error code with wrong height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+  Scenario: To get existing supported networks
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
+	When the user runs the command "query supported-networks"
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To get existing supported networks, Returns an error code with invalid height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario Outline: To get transactions based on the hash, wrong hash
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query the latest upgrade
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query the latest upgrade, invalid height
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Change given, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user has a pocket client
+	When the user runs the command "query node-claims" against validator 0
+	Then the user should be able to see standard output containing "/v1/query/nodeclaims"
+	Then the user should be able to see standard output containing "result"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user has a pocket client
+	When the user runs the command "query nodes" against validator 0
+	Then the user should be able to see standard output containing "/v1/query/nodes"
+	Then the user should be able to see standard output containing "result"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user has a pocket client
+	When the user runs the command "query nodes" against validator 0
+	Then the user should be able to see standard output containing "result"
+	Then the user should be able to see standard output containing "address"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	When the user runs the command "query nodes" against validator 0
+	Then the user should be able to see standard output containing "result"
+	Then the user should be able to see standard output containing "address"
+	Then the user should be able to see standard output containing "jailed"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Then the user should be able to see standard output containing "result"
+	Then the user should be able to see standard output containing "address"
+	Then the user should be able to see standard output containing "jailed"
+	Then the user should be able to see standard output containing "chains"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

ivan-pokt replied:
Seconded! We should not have multiple Then statements following each other.

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Then the user should be able to see standard output containing "address"
+	Then the user should be able to see standard output containing "jailed"
+	Then the user should be able to see standard output containing "chains"
+	Then the user should be able to see standard output containing "service_url"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user has a pocket client
+	When the user runs the command "query param pocketcore/MinimumNumberOfProofs 0" against validator 0
+	Then the user should be able to see standard output containing "param_key"
+	Then the user should be able to see standard output containing "pocketcore/MinimumNumberOfProofs"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user has a pocket client
+	When the user runs the command "query params"
+	Then the user should be able to see standard output containing "param_key"
+	Then the user should be able to see standard output containing "param_value"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user has a pocket client
+	When the user runs the command "query signing-info {{address}} 0" against validator 0
+	Then the user should be able to see standard output containing "/v1/query/signinginfo"
+	Then the user should be able to see standard output containing "result"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user has a pocket client
+	When the user runs the command "query supply"
+	Then the user should be able to see standard output containing "/v1/query/supply"
+	Then the user should be able to see standard output containing "app_staked"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user has a pocket client
+	When the user runs the command "query supported-networks"
+	Then the user should be able to see standard output containing "/v1/query/supportedchains"
+	Then the user should be able to see standard output containing "0011"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user has a pocket client
+	When the user runs the command "query upgrade"
+	Then the user should be able to see standard output containing "/v1/query/upgrade"
+	Then the user should be able to see standard output containing "Height"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
+	When the user runs the command "query upgrade 50000"
+	Then the user should be able to see standard output containing "error loading store"
```
Change Then, by And. Returns the same. Refs: https://www.loom.com/share/81f7d41bcfca480b9f6e9b2ca5ca0651
https://specflow.org/gherkin/five-ways-to-make-gherkin-interesting/
https://docs.behat.org/en/v2.5/guides/1.gherkin.html#scenario-outlines

emedrano2112 commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+	Given the user has a pocket client
+	When the user runs the command "query upgrade 50000"
+	Then the user should be able to see standard output containing "error loading store"
+
```
Missing “exit clause” in all scenarios here (query.feature).

iajrz replied:
Great catch! Adding those.

iajrz committed:
addressing comments about exit status checking and gherkin step keywords

ivan-pokt commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query a non-existing app from address, address has incorrect length but only valid bytes
+	Given the user has a pocket client
+	When the user runs the command "query app 4920ce1d787123456aeff366c79e8aa2 0"
```
Can we possibly make the app ID invalid as well (like giving it a value of 0 or 1 or -1, something impossible to be generated)? We don't want the possibility of a conflict is an app with the same id is created which would make the test fail (given that "4920ce1d787123456aeff366c79e8aa2" refers to the app ID"). Probability of that is low but with more usage, it is more likely - just to be safe.

iajrz replied:
In this scenario the address has an incorrect length, so it's impossible to conflict with a correct app id.

ivan-pokt commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,341 @@
+
+  Scenario: To query dao owner
+	Given the user is running the network "single_node_network"
+	Given the user has a pocket client
```
Let's try to use one Given statement per scenario. We can combine both statements into one Given statement for future usage but ideal to have one Given statement.

ivan-pokt reviewed (changes requested):
I think we need to readjust this to have singular Given statements per scenario and should not have a Then statements immediately following another Then statement. See line comments.

AneudyM commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,391 @@
+	When the user runs the command "query account {{address}}" against validator 0
+	Then the user should be able to see standard output containing "/v1/query/account"
+	And the user should be able to see standard output containing "address"
+	And the pocket client should have exited without error
```
The last And keyword could be replaced with "But should not exit with an error"

iajrz replied:
```suggestion
	But the pocket client should have exited without error
```

iajrz replied:
```suggestion
	But the pocket client should have exited without error
```

AneudyM commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,391 @@
+	Then the user should be able to see standard output containing "Incorrect address length"
+	And the pocket client should have exited without error
+
+  Scenario: To query an existing account, wrong height
```
Scenario titles could be simplified to drive more accuracy and readability. Strive to be consistent with what you have in accounts.feature. It reads more naturally since you have a colon in between the Scenario keyword and the title. For example:

Instead of "Scenario to query an existing account", we could use "Scenario: Query an existing account with wrong height".

iajrz replied:
Titles lifted in the style of the [func tests](https://github.com/pokt-network/pocket-core-func-tests/blob/65fc1cc7a9093b0fcbaab0f52ed3f345827c4de8/tests/cli-commands/pocket-query-commands.feature#L51); will take a look at all scenario titles, thanks for pointing it out.

iajrz replied:
As mentioned elsewhere -- the style was lifted from the original [functional test suite](https://github.com/pokt-network/pocket-core-func-tests/blob/65fc1cc7a9093b0fcbaab0f52ed3f345827c4de8/tests/cli-commands/pocket-query-commands.feature#L51). Will look at all scenario titles.

AneudyM commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,391 @@
+	Then the user should be able to see standard error containing "Usage:"
+	And the pocket client should have exited with error
+
+  Scenario: To query current acl
+	Given the user is running the network "single_node_network"
+	And the user has a pocket client
+	When the user runs the command "query acl"
+	Then the user should be able to see standard output containing "/v1/query/acl"
+	And the user should be able to see standard output containing "gov/non_map_acl"
+	And the pocket client should have exited without error
```
NIT: We can also take advantage of indentation to improve readability. For example, the above scenario could be indented as follows:
```
Scenario: To query current acl
	Given the user is running the network "single_node_network"
	   And the user has a pocket client
	  When the user runs the command "query acl"
	  Then the user should be able to see standard output containing "/v1/query/acl"
	   And the user should be able to see standard output containing "gov/non_map_acl"
	   And the pocket client should have exited without error
```

iajrz replied:
The style of indentation I am using is consistent with the style used by cucumber.io in the [Gherkin language reference ](https://cucumber.io/docs/gherkin/reference/)

AneudyM commented on e2e/tests/query.feature:
```diff
@@ -0,0 +1,391 @@
+	And the user should be able to see standard output containing "0011"
+	And the pocket client should have exited without error
+
+  Scenario: To get existing supported networks, Returns an error code with invalid height
```
Confusing scenario. What is the acceptance criterion here?

Olshansk replied:
@AneudyM I completely with you that this scenario, and it's title are confusing. When reviewing code, I usually try to use Github's "suggestion" feature so the reviewee can just accept/commit it or modify it if need be.; would appreciate your teams help in distributing the workload on this.

For example (I have no clue if this is clearer):
```suggestion
  Scenario: To verify that an invalid height cannot be quried from the list of supported-networks
```

Re the `scenario` - 
1. Are we verifying we can't query an invalid height OR we can't query an unsupported network?

Re the `Then` - `error loading store` makes me think the network is not supported.
2. Is that correct? 
3. If so, is there any way to make the source code (without comments) make that more explicit?

Re the `And`
4. Does this exit successfully because the client worked but the backing data is invalid?

AneudyM replied:
Yeah. Those are good points. Mmm..., looking at the validation of the scenario it might be trying to query a network with an invalid height. In that case, I think I would rewrite it as follows:

````
Scenario: Return error when querying supported networks with invalid height number
  Given a “single_node_network”
    And a valid pocket client
  When querying a supported network with invalid height “99999” number
  Then an “error loading store” message should be returned
    But the pocket client should not exit with an error status
````

Re point 4: 
That's why I am unsure of including the exit status steps here. What does the specification say about these exit statuses.

AneudyM reviewed (changes requested):
The whole query.feature file needs to be refined. It is also too big. Consider using broader-concept steps to avoid repetitions or split them into smaller files if you are unwilling to use background.

iajrz committed:
Update e2e/launcher/node.go

Co-authored-by: Daniel Olshansky <olshansky@pokt.network>

iajrz committed:
Update e2e/tests/query.feature

iajrz committed:
changing fmt for log

Olshansk committed:
rename Readme to README

Olshansk committed:
Minor NIT updates to e2e/launcher/configuration.go to make it easier to read

Olshansk committed:
Automatic VScode updates to .json files

Olshansk committed:
E2E Review

Olshansk committed:
Updated comments

Olshansk commented:
@iajrz I updated the documentation, added a few comments and refactored a bit of the code. I'm in favour of adopting a similar pattern (with a simpler initial implementation) in V1 using gherkin.

Lmk if you see issue with any of my changes. If not, I can go ahead and merge it in.

cc @jessicadaugherty for visibility

Olshansk reviewed (approved):
Got an off-line 👍 from @iajrz on the changes I made so going to go ahead and merge this in:

![Screenshot 2023-01-25 at 12 05 47 PM](https://user-images.githubusercontent.com/1892194/214631366-50f9911a-ba24-4916-8837-5bfce7084a70.png)
